### PR TITLE
Admin Theme Extensibility

### PR DESCRIPTION
**Pull request by Skrypt**
Work In Progress to answer : https://github.com/OrchardCMS/OrchardCore/issues/5605

Using Bootswatch "Darkly" as a base to do a proper admin dark theme at the same time.

orchard

- [x] Allow to extend admin sidebar menu through SASS vars.
- [x] Allow to extend general styles by following practices used in Bootswatch themes.
  - [x] Background headers should use bg-primary instead of bg-light.
- [x] Allow to extend our custom admin styles by defining colors used in SASS vars.
- [ ] Add an option on themes to allow selecting a different stylesheet "bootswatch".
- [ ] Add a user preference to allow selecting a prefered style on a theme. e.g. (dark mode). 
  - [x] LocalStorage based.
  - [ ] Cookie based.
- [ ] Allow extending module styles with admin SASS mixins, funcs, vars.
- [x] Automatically detect dark mode from OS dark mode with javascript. See `prefers-color-scheme: dark` media selector.
- [x] Add admin setting to enable/disable the dark mode admin theme feature.

## Idea

Module extensibility could use a node module that includes all modules SASS files to allow theming them. This way we would be able to use sass variables, functions, mixins to override existing module styles. Right now we have a css generated per module, an alternative would be to move all modules sass files to TheAdminTheme. Using both techniques would allow to properly be able to style every elements from a source code or nuget packages solution AND use SASS.

More on this. We could also add the TheAdmin mixins, func, vars to that node module to allow define our module styles. Right now we can't use the TheAdmin mixins within our modules for example.

### Dark Mode

Tracking Bootstrap Dark Mode feature : twbs/bootstrap#27514

While this should be the way to do this it is more likely to be done for BS 6 rather than 5 so for now we need to rely on a theme switcher for this unfortunately.

Closes : https://github.com/OrchardCMS/OrchardCore/issues/5605

### CODE REVIEW

**Comment by Skrypt** [comment]
I think I did enough for this to get merged. Other features could be achieved in separate PR's. We should merge this sooner than later because of the SASS file structure changes.

File: src/OrchardCore.Themes/TheAdmin/Assets.json
Comment (deanmarcussen) [inline src/OrchardCore.Themes/TheAdmin/Assets.json:6]: Do we have to change the name?

This actually breaks some of my custom admin themes (which rely on the existing name to pull in the normal admin.css code from `TheAdmin` theme, before applying other style overrides)

File: src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs
Comment (deanmarcussen) [inline src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs:14]: This should be `asp-id`.

It also shouldn't be part of the `HtmlTargetElement` attribute set (I think), as this would cause the tag helper to be activated when only using the `asp-id` attribute.

Activation should only happen on the name, src, or at attribute.

**Review by deanmarcussen — COMMENTED** [review]
If I have time I'll have a look at the tag helper later.

I think there might be a nicer way to do that

File: src/OrchardCore.Themes/TheAdmin/Assets.json
Comment (Skrypt) [reply src/OrchardCore.Themes/TheAdmin/Assets.json:6]: Yes because you will be forced to transpile 2 different css now. One for dark mode else the UI will fail to load it. Changing the name forces people to simply reference the new 2 files. In the end you just have to change a name and create a second sass file to generate a second dark mode stylesheet.

File: src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs
Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs:14]: I'm fine with asp-id. I did not create a lot of those TagHelpers I don't know all the tricks yet.

File: src/OrchardCore.Themes/TheAdmin/Assets.json
Comment (Skrypt) [reply src/OrchardCore.Themes/TheAdmin/Assets.json:6]: Plus here, we have a new way to generate those styles without requiring to override on top of the TheAdmin.css file.

**Comment by Skrypt** [comment]
Tracking Bootstrap Dark Mode feature : https://github.com/twbs/bootstrap/issues/27514

File: src/OrchardCore.Themes/TheAdmin/Assets.json
Comment (Skrypt) [reply src/OrchardCore.Themes/TheAdmin/Assets.json:6]: Reverted the name change. Problem solved, changed the javascript to refelct this change.

File: src/OrchardCore.Themes/TheAdmin/Assets/js/darkmode.js
Comment (deanmarcussen) [inline src/OrchardCore.Themes/TheAdmin/Assets/js/darkmode.js:6]: Longer term we will want a way for these user preferences to be stored against the User property rather than in cookies, and for this to be a `Condition` applied to the resource tag. 
But I would not block the feature on this.

**Review by deanmarcussen — COMMENTED** [review]
LGTM. But I think @agriffard pr for the regressions should be merged first, and this css rebuilt.

File: src/OrchardCore.Themes/TheAdmin/Assets/js/darkmode.js
Comment (Skrypt) [reply src/OrchardCore.Themes/TheAdmin/Assets/js/darkmode.js:6]: Last thing I want to add is the automatic selection of the dark mode if your OS is on dark mode.
Will work on small details like these today then I will merge it when I think it's solid enough.

Comment (Skrypt) [reply src/OrchardCore.Themes/TheAdmin/Assets/js/darkmode.js]: I think you meant User IEntity.Properties? Right now I use LocalStorage of the browser which cannot be read from server side. Maybe it should use a Cookie for all those AdminPreferences instead. 

Not sure what you mean with the Condition applied on the resource tag and what it would be for.

File: src/OrchardCore.Themes/TheAdmin/Views/Layout.cshtml
Comment (Skrypt) [inline src/OrchardCore.Themes/TheAdmin/Views/Layout.cshtml:18]: I don't like this. I tried using a @section with a Partial view but seems we can't do that.
Any alternative you guys might think of?

**Comment by Skrypt** [comment]
@deanmarcussen @jtkech Please take a look and tell me if you understand why the CI doesn't pass the functional tests.

**Comment by jtkech** [comment]
@Skrypt Sorry too late i really need to leave, i will take a look tomorrow night if dean didn't fix it in the meantime ;)

File: src/OrchardCore.Modules/OrchardCore.Admin/Admin/AdminSettings.cs
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.Admin/Admin/AdminSettings.cs:7]: I think should be true by default.

Otherwise people will not know it's available

(Same with `DisplayMenuFilter` IMO @agriffard do you agree / disagree?)

File: src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs
Comment (deanmarcussen) [inline src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs:13]: I am flat out today, but I will fix this for you later (before merge), so that any attributes are automatically applied, and we don't need these `asp-id` attributes

**Review by deanmarcussen — COMMENTED** [review]
@Skrypt I took a quick look, and fixed the layout a bit.

The problem with the CI is you've broken both the Setup and Login alternate layouts (they both display blank)

The good news is the mac picked it up automatically (when it was enabled)

File: src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs
Comment (deanmarcussen) [reply src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs]: I tweaked this so that when normal html attributes are supplied to the tag helper it renders them (any of them).

so no special `asp-id` required, just `id` `media` or any other you might need.

File: src/OrchardCore.Themes/TheAdmin/Assets/scss/mixins/_grid.scss
Comment (jptissot) [inline src/OrchardCore.Themes/TheAdmin/Assets/scss/mixins/_grid.scss:21]: eof

File: src/OrchardCore.Themes/TheAdmin/Views/_ViewImports.cshtml
Comment (jptissot) [inline src/OrchardCore.Themes/TheAdmin/Views/_ViewImports.cshtml:12]: eof

**Comment by Skrypt** [comment]
WTF ?? I never asked for a review from @sfmskywalker !!

**Comment by sfmskywalker** [comment]
Someone called? 😁

**Review by sfmskywalker — APPROVED** [review]

**Comment by Skrypt** [comment]
I wish I would have been the one who requested it now I'm asking myself who's been able to use my Github account! 😄

File: src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs
Comment (Skrypt) [reply src/OrchardCore/OrchardCore.ResourceManagement/TagHelpers/StyleTagHelper.cs]: Thank you !

**Comment by jptissot** [comment]
We should just merge my cypress pr ;) that one works :P

**Comment by netwavebe** [comment]
@Skrypt : FYI, just noticed selected tabs (like site settings) are now white text on a white background.

File: gulpfile.js
Comment (deanmarcussen) [inline gulpfile.js:326]: @Skrypt this change has broken building the media app 

or I think any assets that have a comma delimited listing

i.e.
```
    "inputs": [
      "./node_modules/blueimp-file-upload/js/jquery.fileupload.js",
      "./node_modules/blueimp-file-upload/js/jquery.iframe-transport.js",
      "Assets/js/**/*.js"
    ],
```

what's the change for?

Comment (Skrypt) [reply gulpfile.js:326]: To be able to generate a sourceMap file. This is not supposed to be related.

Comment (Skrypt) [reply gulpfile.js:326]: I think someone added a line in the gulpfile.cs so that the Media app be taken into consideration recently.

**Pull request merged**
